### PR TITLE
REF make libboost migrator more robust

### PR DESCRIPTION
**Pull request by beckermr**
- [ ] Pydantic model updated or no update needed

### CODE REVIEW

**Comment by beckermr** [comment]
@h-vetinari I've been seeing a lot of errors in this migrator so I refactored it to be a bit more robust.

File: conda_forge_tick/migrators/libboost.py [hunk 1]
```diff
@@ -58,6 +58,15 @@ def _slice_into_output_sections(meta_yaml_lines, attrs):
     if pos < len(meta_yaml_lines):
         sections[section_index] = meta_yaml_lines[pos:]
 
+    # output_names may contain duplicates; remove them but keep order
```
Comment (beckermr) [inline conda_forge_tick/migrators/libboost.py:61]:
```suggestion
    # double check length here to flag possible weird cases
    # this check will fail incorrectly for outputs with the same name
    # but different build strings.
```

**Comment by beckermr** [comment]
@h-vetinari any comments here?

File: conda_forge_tick/migrators/libboost.py [hunk 2]
```diff
@@ -12,44 +12,106 @@ def _slice_into_output_sections(meta_yaml_lines, attrs):
...
+            and re_output_start.match(line)
+        ):
+            sections[section_index] = meta_yaml_lines[pos:i]
+            section_index += 1
+            pos = i
+            continue
+
+    # the last output possibly needs to be added
+    if pos < len(meta_yaml_lines):
+        sections[section_index] = meta_yaml_lines[pos:]
```
Comment (beckermr) [inline conda_forge_tick/migrators/libboost.py:72]:
```suggestion
    # the last output needs to be added
    sections[section_index] = meta_yaml_lines[pos:]
```

**Comment by beckermr** [comment]
> AFAICT this still doesn't solve templated output-names either. I personally find the index formulation ("index to what?") harder to understand than just identifying a slice by output name, but again, I don't yet understand what problems we're solving here.

It definitely solves the templated output names issue. We no longer require the parser to look for ` name: <name of output>`. The output names are never used except to cut things down to a unique set.

**Comment by beckermr** [comment]
> From looking at the tests, I don't see why these would/should have failed before (aside from the API change of course). What failures did you see? Can we add one of the failing recipes as a test?

Failures are in the bot CI jobs where this code is raising and preventing migrations from going through. I will add a test with those.

File: tests/test_libboost.py [hunk 3]
```diff
@@ -60,3 +61,122 @@ def test_boost(feedstock, new_ver, tmpdir):
...
+        "package:",
+        "  name: blah",
+        "  version: 1.0.0",
+        "",
+        "requirements:",
+        "  host:",
+        "    - foo",
+        "",
+        "outputs:",
+        "  # comment",
```
[the remark this answers is not in the document]
Comment (beckermr) [reply tests/test_libboost.py:126]: Writing these tests was such a pain. This comment would have made it a lot easier!

[the remark this answers is not in the document]
Comment (beckermr) [reply tests/test_libboost.py]: done!

**Comment by h-vetinari** [comment]
> It definitely solves the templated output names issue

OK sorry, that wasn't clear from the test, which is still using hardcoded names. How about having an explicit test à la:
```
outputs:
    - name: {{ name|lower }}
      requirements: [...]
    - name: {{ "lib" ~ name }}
      requirements: [...]
    - name: {{ name ~ "-py" }}
      requirements: [...]
```
There's a bunch of infrastructure for this in `tests/test_libboost.py` where you can put a before/after `meta.yaml` for a given feedstock.

**Comment by beckermr** [comment]
Yep! I am adding some tests per your suggestion here: https://github.com/regro/cf-scripts/pull/2548

They fail with the changes on master and so are marked as xfail. 

The rucio-client one there is interesting. That feedstock failed the slicing but I cannot see why.

**Comment by beckermr** [comment]
> I personally find the index formulation ("index to what?") harder to understand than just identifying a slice by output name, but again, I don't yet understand what problems we're solving here.

I don't disagree here but getting the templated names out of the recipe itself is problematic for a lot reasons including selectors, jinja2 conditional statements, the fact that things like intervening jinja2 statements can potentially change the actual output name even if the template looks the same (mean two unique outputs can have the same template), and that the values of the templated vars are not always easy to deduce.

Counting outputs that come out are ordered seems simpler and a bit more robust as a heuristic, but also has edge cases around jinja2 conditional statements. :(

**Comment by h-vetinari** [comment]
> I don't disagree here but getting the templated names out of the recipe itself is problematic for a lot reasons including selectors

Yeah sure, once we solve that case the extra complexity is very likely justified.

**Comment by beckermr** [comment]
OK. Templated names now pass. There are some unrelated bugs the additional test cases flushed out and there are some actual bugs in the tests themselves. Will push the fixes later.

**Comment by beckermr** [comment]
Alright @h-vetinari, the new tests show the cases the libboost and stdlib migrators can now handle. This will unblock a several failing PRs each bot run, which is a pretty big effect. Templated names for outputs are not rare it turns out.

File: conda_forge_tick/migrators/libboost.py [hunk 4]
```diff
@@ -12,44 +12,105 @@ def _slice_into_output_sections(meta_yaml_lines, attrs):
...
+            pos = i
+            continue
+
+    # the last output needs to be added
+    sections[section_index] = meta_yaml_lines[pos:]
+
+    # finally, if a block list at the same indent happens after the outputs section ends
+    # we'll have extra outputs that are not real. We remove them
+    # by checking if there is a name key in the dict
+    re_name = re.compile(r"^\s*(-\s+)?name:.*")
```
Comment (h-vetinari) [inline conda_forge_tick/migrators/libboost.py:76]: I don't get the effort here. Why don't we just match for `- name: ...` directly, rather than determining the indent level, catching everything at that indent level, and _then_ matching on name?

File: conda_forge_tick/migrators/libboost.py [hunk 5]
```diff
@@ -12,44 +12,105 @@ def _slice_into_output_sections(meta_yaml_lines, attrs):
...
+                )
+
+            outputs_token_pos = i
+            continue
+
+        # next we need to find the indent of the first output list element.
+        # it could be anything from the indent of the `outputs:` line to that number
+        # plus zero or more spaces. Typically, the indent is the same as the `outputs:`
+        # plus one of [0, 2, 4] spaces.
+        # ocne we find the first output list element, we build a regex to match any list
```
Comment (h-vetinari) [inline conda_forge_tick/migrators/libboost.py:46]: ocne -> once

(though I'd like to avoid the indentation handling completely - I don't (yet?) see why it's necessary)

**Review by h-vetinari — COMMENTED** [review]
some questions about the implementation

File: conda_forge_tick/migrators/libboost.py (hunk 4 above)
Comment (beckermr) [reply conda_forge_tick/migrators/libboost.py:76]: An output doesn't have to start with the name key as the first item in the dictionary. People do this as a custom but it is not required.

File: conda_forge_tick/migrators/libboost.py (hunk 5 above)
Comment (beckermr) [reply conda_forge_tick/migrators/libboost.py:46]: Thanks. I will fix the typo. 

The indentation handling is because yaml allows any consistent indent below a given section. The outputs are a list, is you can have

"""
outputs:
- blah
"""

Or 

"""
outputs:
  - blah
"""

Or w/e

Comment (beckermr) [inline conda_forge_tick/migrators/libboost.py:46]:
```suggestion
        # once we find the first output list element, we build a regex to match any list
```

Comment (beckermr) [reply conda_forge_tick/migrators/libboost.py]: Looking for list items at any indent is not robust ofc, so we have to find a list item at the correct indent. Even that is not robust since we don't know when the outputs section ends. Thus we clean that up later by requiring a `name: ` key in the output. This is a heuristic since in principle that string could appear in other parts of the recipe.

Comment (beckermr) [reply conda_forge_tick/migrators/libboost.py]: We could detect the indent of the dict in each output, record it, and then find the right name key at the correct indent.

File: conda_forge_tick/migrators/libboost.py (hunk 4 above)
Comment (h-vetinari) [reply conda_forge_tick/migrators/libboost.py:76]: > An output doesn't have to start with the name key as the first item in the dictionary.

Thanks, I hate it. 🙈 

I've _never_ seen that though (in contrast to templated outputs). In fact, I'd _want_ such feedstocks to break, because doing
```
outputs:
  - build:
      skip: true  # [not linux]
    name: ...
```
should just not be supported at all (i.e. we should turn it into a hard linter error) IMO.

I certainly don't think it's worth the implementation effort to accommodate such an extreme corner case.

File: conda_forge_tick/migrators/libboost.py (hunk 5 above)
Comment (h-vetinari) [reply conda_forge_tick/migrators/libboost.py]: Personally, I just want to not care about the indent at all here, and let the regex portion of `\s*` take care of it.

File: conda_forge_tick/migrators/libboost.py (hunk 4 above)
Comment (beckermr) [reply conda_forge_tick/migrators/libboost.py:76]: Yeah right now the libboost and cstdlib migrators are blocking perfectly good version and migration prs from being made because they don't handle templated names. 

We have one of these options

- turn off these minimigrators
- swallow their errors so they don't block PRs
- merge this pr to help fix things
- merge another pr to help fix things
- go around telling people the bot cannot handle templated output names

The implementation effort is fundamentally what is required to do the job at some standard, not what we are willing to give. The current recipe format allows a lot of edge cases and that generates complicated code in the bot. That's just how it is.

File: conda_forge_tick/migrators/libboost.py (hunk 5 above)
Comment (beckermr) [reply conda_forge_tick/migrators/libboost.py]: That's less robust and leaves edge cases.

**Comment by beckermr** [comment]
@h-vetinari I don't get the resistance to this PR. 

If it makes you feel better, you can defer all further issues with this code or the associated bot aspects of those migrations to me. That way you don't have any extra work associated with code you feel is too complex. 

The current code misses too many edge cases and that has to be fixed.

File: conda_forge_tick/migrators/libboost.py (hunk 4 above)
Comment (h-vetinari) [reply conda_forge_tick/migrators/libboost.py:76]: There was a misunderstanding here. I agree we should fix templated names, no question about that (even though I'd like to get away from them wherever possible).

What I meant is that I don't think we should fix "name of output does not appear as first line", because the implementation here has IMO ballooned way further than what it is necessary to support templated outputs - which would mainly be to relax `f"- name: {seek}"` -> `f"- name:` (i.e. not include the package name), and incorporate your array/index logic of course.

If you feel the added generality for catching this extreme corner case[^1] is worth it, I won't stand in the way, but it makes me way less likely to touch that code again because I feel it's a bad trade-off and would have to resist the urge to remedy that.

[^1]: again, not talking about templated names here, but about not relying on having the hyphen before `name:`, i.e. `name:` being the first line per output

Comment (beckermr) [reply conda_forge_tick/migrators/libboost.py:76]: That's fine. I'll take care of this code from now on. Thanks for the help in getting it started!

**Comment by h-vetinari** [comment]
> @h-vetinari I don't get the resistance to this PR.

The implementation complexity makes a big jump, I'm simply asking if it's worth it in all aspects. That's far from resistance (I said I won't stand in the way), just regular review process.

> The current code misses too many edge cases and that has to be fixed.

You should IMO add tests for all the cases you intend the implementation to cover, otherwise it's too easy for things to regress.

File: conda_forge_tick/migrators/libboost.py (hunk 4 above)
Comment (h-vetinari) [reply conda_forge_tick/migrators/libboost.py:76]: We might also want to move this to be an "official" (albeit smith-internal) utility, rather than being tucked away in the boost migrator.

**Comment by beckermr** [comment]
Good point! Both the templated names and outputs that don't have name as the first dict item are in the tests here already.

**Comment by h-vetinari** [comment]
> outputs that don't have name as the first dict item are in the tests here already.

I disagree. You have
```
outputs:
  # comment
    - name: blarg  # comment
      requirements:
```
but that's just a comment in between (same goes for the cases from #2548). IOW, every `name:` field still actually comes with `- ` in front (and so it would have been sufficient to detect section boundaries by `\s*-\s+name:.*`, at least for for the current test corpus). If your intention was to support something like
```
outputs:
  - build:
      skip: true  # [not linux]
    name: ...
```
then I suggest you add a test for that.

**Comment by h-vetinari** [comment]
Nevermind, I just now saw
```
    - requirements:
        host:
          - baz
      name: blarg2
```
Sorry about the noise.
